### PR TITLE
[mlir][VectorOps] Extend vector.constant_mask to support 'all true' scalable dims

### PR DESCRIPTION
**Pull request by MacDue**
This extends `vector.constant_mask` so that mask dim sizes that correspond to a scalable dimension are treated as if they're implicitly multiplied by vscale. Currently this is limited to mask dim sizes of 0 or the size of the dim/vscale. This allows constant masks to represent all true and all false scalable masks (and some variations):

```
// All true scalable mask
%mask = vector.constant_mask [8] : vector<[8]xi1>

// All false scalable mask
%mask = vector.constant_mask [0] : vector<[8]xi1>

// First two scalable rows
%mask = vector.constant_mask [2,4] : vector<4x[4]xi1>
```

### CODE REVIEW

**Comment by dcaballe** [comment]
THanks for the contribution. Really interesting... Is the constant input required to be always lower or equal to the scalable base?

For example, what these two examples would mean:

```
%mask = vector.constant_mask [2, 7] : vector<4x[4]xi1>
%mask = vector.constant_mask [2, 3] : vector<4x[4]xi1>
```

I'm also wondering if it would make sense to distinguish between `[2, 4] : vector<4x[4]xi1>` and `[2, [4]] : vector<4x[4]xi1>` as they may mean completely different things?

**Comment by banach-space** [comment]
> Is the constant input required to be always lower or equal to the scalable base?

ATM it is required to be "equal". 

> 
> For example, what these two examples would mean:
> 
> ```
> %mask = vector.constant_mask [2, 7] : vector<4x[4]xi1>
> %mask = vector.constant_mask [2, 3] : vector<4x[4]xi1>
> ```

Indeed, it's unclear and hence I suggest that it's not supported until we figure this out :)

> 
> I'm also wondering if it would make sense to distinguish between `[2, 4] : vector<4x[4]xi1>` and `[2, [4]] : vector<4x[4]xi1>` as they may mean completely different things?

I can't think of a use case for this that couldn't be accommodated with `vector.create_mask`. 

For cases like this one [here](https://github.com/llvm/llvm-project/pull/65621#discussion_r1321037163), it should be sufficient if we have a Vector op that we can use to create an "all true" mask for scalable vectors (instead of using`arith.constant dense<true> : vector<[8]xi1>`). IIUC, this is what this patch  strives to achieve. As for more complex cases, we should get a better idea in the coming weeks/months as we continue extending/refining support for SVE and scalable vectors.

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorMask.cpp [hunk 1]
```diff
@@ -115,43 +114,41 @@ class ConstantMaskOpLowering : public OpRewritePattern<vector::ConstantMaskOp> {
...
-          op, dstType, rewriter.getBoolVectorAttr(values));
+      if (trueDimSize == 0 || trueDimSize == dstType.getDimSize(0)) {
+        // Use constant splat for 'all set' or 'none set' dims.
+        // This produces correct code for scalable dimensions.
+        rewriter.replaceOpWithNewOp<arith::ConstantOp>(
+            op, DenseElementsAttr::get(dstType, trueDimSize != 0));
+      } else {
+        // Express constant 1-D case in explicit vector form:
+        //   [T,..,T,F,..,F].
+        SmallVector<bool> values(dstType.getDimSize(0));
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorMask.cpp:133]: adding `false` value would help readability

**Review by dcaballe — APPROVED** [review]
Thanks for the clarification. LGTM!

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir [hunk 2]
```diff
@@ -1819,16 +1819,55 @@ func.func @genbool_1d() -> vector<8xi1> {
...
+  %0 = vector.constant_mask [8] : vector<[8]xi1>
+  return %0 : vector<[8]xi1>
+}
+// CHECK-LABEL: func @genbool_1d_scalable_ptrue
+// CHECK: %[[VAL_0:.*]] = arith.constant dense<true> : vector<[8]xi1>
+// CHECK: return %[[VAL_0]] : vector<[8]xi1>
+
+// -----
+
+func.func @genbool_2d_scalable() -> vector<4x[4]xi1> {
```
Comment (c-rhodes) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:1842]: `genbool_2d_trailing_scalable`

File: mlir/test/Dialect/Vector/ops.mlir [hunk 3]
```diff
@@ -1003,7 +1007,7 @@ func.func @contraction_masked_scalable(%A: vector<3x4xf32>,
                                     %C: vector<3x[8]xf32>,
                                     %M : vector<3x[8]x4xi1>) -> vector<3x[8]xf32> {
  // CHECK:  vector.mask %[[M]] { vector.contract {indexing_maps = [#{{.*}}, #{{.*}}, #{{.*}}], iterator_types = ["parallel", "parallel", "reduction"], kind = #vector.kind<add>} %[[A]], %[[B]], %[[C]] : vector<3x4xf32>, vector<4x[8]xf32> into vector<3x[8]xf32> } : vector<3x[8]x4xi1> -> vector<3x[8]xf32>
-  %0 = vector.mask %M { vector.contract #matmat_trait %A, %B, %C : vector<3x4xf32>, vector<4x[8]xf32> into vector<3x[8]xf32> } 
+  %0 = vector.mask %M { vector.contract #matmat_trait %A, %B, %C : vector<3x4xf32>, vector<4x[8]xf32> into vector<3x[8]xf32> }
```
Comment (c-rhodes) [inline mlir/test/Dialect/Vector/ops.mlir:1010]: unrelated change?

**Review by c-rhodes — APPROVED** [review]
Thanks Ben couple of minor comments but otherwise LGTM

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir [hunk 4]
```diff
@@ -1819,16 +1819,55 @@ func.func @genbool_1d() -> vector<8xi1> {
 
 // -----
 
-func.func @genbool_1d_scalable() -> vector<[8]xi1> {
```
Comment (banach-space) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:1822]: [nit] `ptrue` is a very "SVE term", not sure about `pfalse` tbh. Perhaps `all_true` and `all_false` instead? That would be more generic.

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir [hunk 5]
```diff
@@ -1819,16 +1819,55 @@ func.func @genbool_1d() -> vector<8xi1> {
...
+}
+// CHECK-LABEL:   func.func @genbool_2d_scalable() -> vector<4x[4]xi1> {
+// CHECK:           %[[VAL_0:.*]] = arith.constant dense<true> : vector<[4]xi1>
+// CHECK:           %[[VAL_1:.*]] = arith.constant dense<false> : vector<4x[4]xi1>
+// CHECK:           %[[VAL_2:.*]] = builtin.unrealized_conversion_cast %[[VAL_1]] : vector<4x[4]xi1> to !llvm.array<4 x vector<[4]xi1>>
+// CHECK:           %[[VAL_3:.*]] = llvm.insertvalue %[[VAL_0]], %[[VAL_2]][0] : !llvm.array<4 x vector<[4]xi1>>
+// CHECK:           %[[VAL_4:.*]] = llvm.insertvalue %[[VAL_0]], %[[VAL_3]][1] : !llvm.array<4 x vector<[4]xi1>>
+// CHECK:           %[[VAL_5:.*]] = builtin.unrealized_conversion_cast %[[VAL_4]] : !llvm.array<4 x vector<[4]xi1>> to vector<4x[4]xi1>
+// CHECK:           return %[[VAL_5]] : vector<4x[4]xi1>
+// CHECK:         }
```
Comment (banach-space) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:1854]: DELETEME (not needed)

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir [hunk 6]
```diff
@@ -1819,16 +1819,55 @@ func.func @genbool_1d() -> vector<8xi1> {
...
+/// Currently, this is not supported as generating the mask would require
+/// unrolling the leading scalable dimension at compile time.
+func.func @cannot_genbool_2d_leading_scalable() -> vector<[4]x4xi1> {
+  %0 = vector.constant_mask [4, 2] : vector<[4]x4xi1>
+  return %0 : vector<[4]x4xi1>
+}
+// CHECK-LABEL:   func.func @cannot_genbool_2d_leading_scalable() -> vector<[4]x4xi1> {
+// CHECK:           %[[VAL_0:.*]] = vector.constant_mask [4, 2] : vector<[4]x4xi1>
+// CHECK:           return %[[VAL_0]] : vector<[4]x4xi1>
+// CHECK:         }
```
Comment (banach-space) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:1867]: DELETEME (not needed)

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorMask.cpp [hunk 7]
```diff
@@ -115,43 +114,41 @@ class ConstantMaskOpLowering : public OpRewritePattern<vector::ConstantMaskOp> {
...
     }
 
-    VectorType lowType =
-        VectorType::get(dstType.getShape().drop_front(), eltType);
-    SmallVector<int64_t> newDimSizes;
-    for (int64_t r = 1; r < rank; r++)
-      newDimSizes.push_back(cast<IntegerAttr>(dimSizes[r]).getInt());
+    if (dstType.getScalableDims().front())
+      return rewriter.notifyMatchFailure(
+          op, "Cannot unroll leading scalable dim in dstType");
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorMask.cpp:144]: We should be able to test for this in invalid.mlir, right?

**Review by banach-space — APPROVED** [review]
LGTM (% minor suggestions), thanks!

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorMask.cpp (hunk 7 above)
Comment (MacDue) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorMask.cpp:144]: No, this is a match failure not an invalid op (this is just a fancy `return failure()`), it does not produce a diagnostic.

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir (hunk 5 above)
Comment (MacDue) [reply mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir]: I take this to mean the `CHECK: }` , was a little confused by this comment.

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir (hunk 6 above)
Comment (MacDue) [reply mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir]: I take this to mean the `CHECK: }`  
Was a little confused by this comment.

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorMask.cpp [hunk 8]
```diff
@@ -115,43 +114,41 @@ class ConstantMaskOpLowering : public OpRewritePattern<vector::ConstantMaskOp> {
...
-      // Express constant 1-D case in explicit vector form:
-      //   [T,..,T,F,..,F].
-      SmallVector<bool> values(dstType.getDimSize(0));
-      for (int64_t d = 0; d < trueDim; d++)
-        values[d] = true;
-      rewriter.replaceOpWithNewOp<arith::ConstantOp>(
-          op, dstType, rewriter.getBoolVectorAttr(values));
+      if (trueDimSize == 0 || trueDimSize == dstType.getDimSize(0)) {
+        // Use constant splat for 'all set' or 'none set' dims.
+        // This produces correct code for scalable dimensions.
```
[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorMask.cpp]: Yep, it becomes a constant splat (noted this in the comment now)

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir (hunk 6 above)
Comment (banach-space) [reply mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir]: Yes, meant that final `CHECK`. Sorry, Phabricator would highlight the file line that the comment was meant to be attached to :/
